### PR TITLE
memtable supports dynamic memory allocation

### PR DESCRIPTION
**Pull request by hslam**

### CODE REVIEW

File: table/memtable/arena.go [hunk 1]
```diff
@@ -154,16 +168,22 @@ func (vn *valueNode) decode(b []byte) {
...
+	vl.decode(s.get(addr))
 	return vl
 }
+
+func markValueNodeAddr(addr *arenaAddr) {
+	*addr = arenaAddr(1<<valueNodeShift | uint64(*addr))
+}
+
+func isValueNodeAddr(addr arenaAddr) bool {
+	return uint64(addr)&valueNodeMask>>valueNodeShift == 1
```
Comment (coocood) [inline table/memtable/arena.go:188]: `uint64(addr)&valueNodeMask != 0`

File: table/memtable/arena.go [hunk 2]
```diff
@@ -25,113 +25,127 @@ import (
 )
 
 const (
-	offsetSize = int(unsafe.Sizeof(uint32(0)))
+	offsetSize = int(unsafe.Sizeof(uint64(0)))
```
Comment (coocood) [inline table/memtable/arena.go:28]: `offsetSize = int(unsafe.Sizeof(arenaAddr(0)))`

File: table/memtable/arena_locator.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+		return a.buf[offset : offset+uint32(size)]
+	}
+	return a.buf[offset:]
+}
+
+func (a *arenaBlock) alloc(size int) uint32 {
+	// The returned addr should be aligned in 8 bytes.
+	offset := (a.length + blockAlign) & alignMask
+	a.length = offset + size
+	if a.length > len(a.buf) {
```
Comment (coocood) [inline table/memtable/arena_locator.go:131]: Better check and return before update `a.length`

File: table/memtable/arena.go (hunk 1 above)
Comment (hslam) [reply table/memtable/arena.go]: fixed

File: table/memtable/arena.go (hunk 2 above)
Comment (hslam) [reply table/memtable/arena.go]: fixed

File: table/memtable/arena_locator.go (hunk 3 above)
Comment (hslam) [reply table/memtable/arena_locator.go]: fixed

**Comment by coocood** [comment]
LGTM

**Pull request merged**